### ISSUE TITLE
How to safely kill a in flight process

### ISSUE THREAD

**Issue by kalepail** [RReverser/wasm-bindgen-rayon#19]
Forgive me if this question is basic or out of scope, I'm fairly new to Rust and WASM.

Is there a mechanic for killing an inflight request? I have a program that takes some arguments and could begin taking a long time to process. It's possible I may want to kill an inflight request, change some arguments, and try again without waiting for the current request to complete. Is there a mechanic for ensuring this is done safely?

Thanks! Really appreciate this package and the work put into it. I was using a cobbled together version of this https://wasm-bindgen.netlify.app/examples/raytrace previously. This is so much cleaner!

**Comment by RReverser** [comment]
Sorry, I don't really know what "killing an inflight request" means in context of this library. Can you show an example?

**Comment by kalepail** [comment]
You know how this demo allows you to make requests during the process https://rustwasm.github.io/wasm-bindgen/examples/raytrace.html (get image progress). I can't get a similar function to work with this demo. The call locks up until it resolves. This means I can't reach into the request to do things like "get image progress" or "kill the process while still returning the latest data". Am I missing something? Seems odd the request locks up. How is that raytracing demo able to do it but this one isn't?

**Comment by RReverser** [comment]
That's because they're running the demo code in a Web Worker. It has to do with this limitation mentioned in the docs:

> The main thread in a browser cannot block. This means that if you run WebAssembly code on the main thread you can never block, meaning you can't do so much as acquire a mutex. This is an extremely difficult limitation to work with on the web, although one workaround is to run Wasm exclusively in web workers and run JS on the main thread. 

We used to have the same limitation in wasm-bindgen-rayon, but it proved too cumbersome to work with, since you need to manage that first Worker yourself, send messages to it manually, receive messages from it to update UI and so on. That amount of code rearchitecturing is just way too much of painful work when you simply want to speed up a small function by parallelising it.

Instead, nowadays we have a workaround for that issue that spin-locks on the main thread. This way, code can work on the main thread with no changes whether parallelised or not, but users who need to run longer-running operations, still get an option to create their own Workers.

TL;DR - if you need to run long-running operations, create your own Worker and run said code over there asynchronously.

**Comment by RReverser** [comment]
Going to close this as I hope the question has been answered.
